### PR TITLE
Additions to FallingFactorial

### PR DESCRIPTION
**Pull request by leosartaj**
### TODO
- [X] Evaluate `ff(n, n)` to `factorial(n)`.
- [X] Rewite in terms of `rf`.
- [X] Rewrite `ff` in terms of `binomial`.
- [X] Rewrite `ff` in terms of `factorial`.

Can you please review this? @jksuom

### CODE REVIEW

File: sympy/functions/combinatorial/factorials.py [hunk 1]
```diff
@@ -541,6 +561,8 @@ def eval(cls, x, k):
 
         if x is S.NaN:
             return S.NaN
+        elif x == k:
+            return factorial(x)
```
Comment (jksuom) [inline sympy/functions/combinatorial/factorials.py]: I think ``factorial`` should be returned only if ``x == k`` is integer.

File: sympy/functions/combinatorial/factorials.py [hunk 2]
```diff
@@ -541,6 +561,8 @@ def eval(cls, x, k):
...
         elif k.is_Integer:
             if k is S.NaN:
```
Comment (jksuom) [inline sympy/functions/combinatorial/factorials.py]: ``k.is_Integer`` and ``k is S.NaN``. Is this really possible?

**Comment by jksuom** [comment]
These are a couple of quick comments before I have to shut down for the weekend. I'll be back in a couple of days.

File: sympy/functions/combinatorial/factorials.py (hunk 2 above)
Comment (leosartaj) [inline sympy/functions/combinatorial/factorials.py]: That's not possible. I will go through ``eval`` method. This should have been caught. Probably this is not very well tested.

File: sympy/functions/combinatorial/factorials.py (hunk 1 above)
Comment (leosartaj) [inline sympy/functions/combinatorial/factorials.py]: That would be better.

**Comment by jksuom** [comment]
I feel uneasy about the ``fdiff`` method. The falling factorial is a polynomial of its first argument, so I would expect its derivative to be a polynomial, too. The second argument should be restricted to integer values, and derivation should lead to an error. If a user really wants to have a derivative involving transcendental functions, he can rewrite ``ff`` in terms of the gamma function before derivation. (So maybe ``ff.fdiff`` is not necessary at all.)

**Comment by leosartaj** [comment]
``ff.fdiff`` result is similar what to WolframAlpha gives. How about first converting to ``gamma`` and then differentiating? I am not very comfortable about raising an error, when derivation is actually defined.

**Comment by jksuom** [comment]
I would very much like to keep the Pochhammer symbols and their derivatives rational. Although they can be expressed in terms of ``gamma`` and ``polygamma``, I don't think such conversions should be automatic. (And I am not sure we should follow every aspect of WolframAlpha.)

I have doubts about the usefulness of these derivatives in general, but one possibility might be an explicit computation as a polynomial along the following lines: The derivative of the product ``x(x - 1)...(x - k + 1)`` is the sum of the products ``x(x - 1)...(x - i + 1)(x - i - 1)...(x - k + 1)``, and these can again be written as products of two ``ff``'s.

**Comment by leosartaj** [comment]
I have got rid of ``fdiff`` @jksuom.

**Comment by jksuom** [comment]
That is probably reasonable. (The rational expression I suggested is, in fact, quite complicated.)

There is another thing: I would like to restrict the use of ``factorial`` to integral arguments. (Else ``gamma`` can be used.) So perhaps there should be a test in ``_eval_rewrite_as_factorial``. (``k`` should be tested already when ``ff(x, k)`` is constructed.)

Otherwise, I think this looks good.

**Comment by leosartaj** [comment]
Should we also restrict rewriting to binomial only when arguments are integral?

**Comment by jksuom** [comment]
I think it would suffice for the second argument to be integral. Then can we have ``binomial(x, k) = ff(x, k)/factorial(k)`` with no restriction on ``x`` in either ``ff`` or ``binomial``. Both will be rational expressions of ``x`` for integral values of ``k``.

**Comment by leosartaj** [comment]
Please have a look @jksuom.

**Comment by jksuom** [comment]
Looks good. +1

**Comment by jksuom** [comment]
@leosartaj I think this is ready unless you wish to squash some commits.

**Comment by leosartaj** [comment]
I have squashed the commits @jksuom

**Comment by jksuom** [comment]
Thanks. Merging.

**Comment by leosartaj** [comment]
Thanks for the review @jksuom.

**Pull request merged**